### PR TITLE
Support for an IPluginStorageFilter

### PR DESCRIPTION
**Pull request by mlucool**
Same as previous two commits - just not commited to master

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -437,4 +438,8 @@ type StringValue = string | void | null;
 	interface IPluginMiddleware<T> extends IPlugin<T> {
 		register_middlewares(app: any, auth: IBasicAuth<T>, storage: IStorageManager<T>): void;
 	}
+
+	interface IPluginStorageFilter<T> extends IPlugin<T> {
+		filter(packageInfo: Package, cb: Callback): void;
```
Comment (juanpicado) [inline index.d.ts]: Here you are filtering `metadata` only. But the method name is too generic, perhaps someone wanna filter `distTags`, `versions`, or avoid `removePackages` in the future. I think this should be more specific to the task is executing, in this case filtering metadata.

Comment (mlucool) [reply index.d.ts]: `filter_metadata`?

Comment (juanpicado) [reply index.d.ts]: go ahead, LGTM

**Comment by mlucool** [comment]
This is ready to be merged pending your review. The `close` addition is required as a plugin needs to be able to be reset in testing. Maybe one day verdaccio may want on the fly starting/closing of plugins.

File: src/flow/types.js [hunk 2]
```diff
@@ -406,6 +409,7 @@ declare interface verdaccio$IBasicAuth {
 
 declare interface verdaccio$IPlugin {
   version?: string;
+  close(): void;
```
Comment (juanpicado) [inline src/flow/types.js]: This should be optional, otherwise will force current plugins to implement it.

**Pull request merged**